### PR TITLE
Story/hfi ready toggle/1859

### PR DESCRIPTION
**Pull request by conbrad**
**_Given I an using the HFI Calc with Permissions to edit Planning Area (R/ZWCO):_**
 - [x] When I load the product on prep morning, all planning areas are in a 'not ready' toggle state
     - `ToggleOffOutlinedIcon`
 - [x] When I am ready, I can manually change the status of a single planning area to 'ready' by clicking the toggle
      - `ToggleOnOutlinedIcon` 
 - [x] When a planning area has been marked as ready, there is a hover state that says: Marked as ready by [IDIR] on [timestamp]
 - [x] When a user makes a change to that planning area after marking it ready (include/exclude, fuel, fire starts) the status is auto-reverted back to not ready
 - [x] When a planning area has been marked as ready or is auto-reverted, a success toast is generated

**_Given I an using the HFI Calc withOUT Permissions to edit Planning Area (read only or other role):_**
- [x] When I visit the product, all toggles are visible but in a disabled state - I cannot click them and this is clear to me
- [x]  When I visit the product, the hover state on 'ready' planning areas is available to me
# Test Links:
[Percentile Calculator](https://wps-pr-2087.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-2087.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-2087.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-2087.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-2087.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Fire Behaviour Advisory](https://wps-pr-2087.apps.silver.devops.gov.bc.ca/fire-behaviour-advisory)
[HFI Calculator](https://wps-pr-2087.apps.silver.devops.gov.bc.ca/hfi-calculator)
[FWI Calculator](https://wps-pr-2087.apps.silver.devops.gov.bc.ca/fwi-calculator)

### CODE REVIEW

**Review by tessmgood — CHANGES_REQUESTED** [review]
AC: When a user makes a change to that planning area after marking it ready (include/exclude, fuel, fire starts) the status is auto-reverted back to not ready

> can we revert back to not ready for THAT planning area only? presently all planning areas revert back to not ready, which might undo someone else's work

**Comment by tessmgood** [comment]
AC: When a planning area has been marked as ready or is auto-reverted, a success toast is generated

I'm not getting a success toast when I mark as ready

**Comment by tessmgood** [comment]
AC: When I load the product on prep morning, all planning areas are in a 'not ready' toggle state

The toggle is disabled until I interact with the page - good instinct, assumption is that something should change, BUT it might not, e.g. anticipated fire starts might be low (its cool/raining) and all the data is in (no need to exclude) = no adjustments needed. In this case, it's not apparent what to do / why the toggle is disabled. 

TL;DR can we enable the toggle on first load? If they are silly and mark as ready, then start making changes, our auto-revert feature will set them straight

File: api/app/db/crud/hfi_calc.py [hunk 1]
```diff
@@ -145,6 +145,28 @@ def store_hfi_station(session: Session, station_code: int, fuel_type_id: int, pl
     session.commit()
 
 
+def toggle_ready(session: Session,
+                 planning_area_id: int,
+                 hfi_request_id: int,
+                 username: str) -> HFIReady:
+    """ Toggles the planning area ready state for an hfi request """
+    now = get_pst_now()
```
Comment (Sybrand) [inline api/app/db/crud/hfi_calc.py:153]: I'd prefer we conform to using utc wherever possible.

File: api/app/tests/hfi/test_hfi_ready.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    "planning_area_hfi_results": [],
+    "planning_area_fire_starts": {}
+}
+
+ready_state_json = {
+    "hfi_request_id": 1,
+    "planning_area_id": 1,
+    "ready": False,
+    "create_user": 'test',
+    "create_timestamp": datetime.fromtimestamp(1560130969).isoformat(),
```
Comment (Sybrand) [inline api/app/tests/hfi/test_hfi_ready.py:31]: Easier to read?:
```suggestion
    'create_timestamp': '2019-06-09T18:42:49'
```

File: api/app/db/crud/hfi_calc.py [hunk 3]
```diff
@@ -145,6 +145,28 @@ def store_hfi_station(session: Session, station_code: int, fuel_type_id: int, pl
...
+    ready_state: HFIReady = session.query(HFIReady)\
+        .filter(HFIReady.planning_area_id == planning_area_id)\
+        .filter(HFIReady.hfi_request_id == hfi_request_id)\
+        .first()
+    if ready_state.ready is True:
+        ready_state.ready = False
+    else:
+        ready_state.ready = True
+    ready_state.update_timestamp = now
+    ready_state.update_user = username
```
Comment (andrea-williams) [inline api/app/db/crud/hfi_calc.py:163]: Was just playing around with this in dev. Having the tooltip msg with username & timestamp of who marked it ready is a real nice detail!

File: api/app/tests/hfi/test_hfi_ready.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    "update_user": 'test',
+    "update_timestamp": datetime.fromtimestamp(1560217369).isoformat()
```
Comment (Sybrand) [inline api/app/tests/hfi/test_hfi_ready.py:33]: Easier to read?
```suggestion
    "update_timestamp": '2019-06-10T18:42:49'
```

File: api/app/schemas/hfi_calc.py [hunk 5]
```diff
@@ -197,6 +198,20 @@ class HFIResultResponse(BaseModel):
     fire_start_ranges: List[FireStartRange]
 
 
+class HFIReadyState(BaseModel):
+    planning_area_id: int
+    hfi_request_id: int
+    ready: bool
+    create_timestamp: datetime
+    create_user: str
+    update_timestamp: datetime
```
Comment (andrea-williams) [inline api/app/schemas/hfi_calc.py:207]: does the ready state need to distinguish between create and update? When are they used differently? (Or are you just following standard practice for DBs?)

**Review by Sybrand — CHANGES_REQUESTED** [review]
Changing fire starts doesn't change the ready state on the screen until I refresh.

**Review by andrea-williams — CHANGES_REQUESTED** [review]
If a planning area is already marked as Ready and then I make a change to the fire starts value, the ready toggle isn't immediately updated, it's only updated if I refresh the page after making the change. So to the user making the change, it looks like the area is marked ready, but then if my boss comes along later, it looks like I haven't done my work yet.

File: api/app/schemas/hfi_calc.py (hunk 5 above)
Comment (conbrad) [reply api/app/schemas/hfi_calc.py:207]: There will be a difference -- whoever makes the first change to a prep period will create the first ready record, then all other changes will be updates. Technically we mark `create_timestamp` and `update_timestamp` to be the same on creation, but yeah following standard practice there.

**Review by tessmgood — COMMENTED** [review]
When a planning area has been marked as ready or is auto-reverted, a success toast is generated
- not getting success toasts at all on the toggle 
- I got a success toast the first time I changed a fire starts dropdown, all changes after that did not generate a toast

File: web/src/api/hfiCalculatorAPI.ts [hunk 6]
```diff
@@ -145,8 +146,8 @@ export async function getAllReadyStates(
   const { data } = await axios.get<HFIAllReadyStatesResponse>(url)
   return data.ready_states.map(planningAreaReadyState => ({
     ...planningAreaReadyState,
-    create_timestamp: DateTime.fromISO(planningAreaReadyState.create_timestamp),
-    update_timestamp: DateTime.fromISO(planningAreaReadyState.update_timestamp)
+    create_timestamp: DateTime.fromISO(planningAreaReadyState.create_timestamp).setZone(PACIFIC_IANA_TIMEZONE),
```
[the remark this answers is not in the document]
Comment (conbrad) [reply web/src/api/hfiCalculatorAPI.ts:149]: No, we've discussed this before and agreed that the API is the place to handle all date changes until we implement our wrapper class around dates.

**Review by Sybrand — APPROVED** [review]

**Review by tessmgood — APPROVED** [review]
Niceeeeeeee

File: api/app/db/crud/hfi_calc.py [hunk 7]
```diff
@@ -145,6 +145,29 @@ def store_hfi_station(session: Session, station_code: int, fuel_type_id: int, pl
...
+    now = get_utc_now()
+    hfi_request = get_most_recent_updated_hfi_request(session, fire_centre_id, date_range)
+    ready_state: HFIReady = session.query(HFIReady)\
+        .filter(HFIReady.planning_area_id == planning_area_id)\
+        .filter(HFIReady.hfi_request_id == hfi_request.id)\
+        .first()
+    if ready_state.ready is True:
+        ready_state.ready = False
+    else:
+        ready_state.ready = True
```
Comment (andrea-williams) [inline api/app/db/crud/hfi_calc.py:163]: couldn't you replace these last 4 lines with just `ready_state.ready = !ready_state.ready` ?

Comment (conbrad) [reply api/app/db/crud/hfi_calc.py:163]: Yeah nice call. I can do that in a follow up PR.
